### PR TITLE
app-editors/scite:

### PR DESCRIPTION
**Pull request by dermoench42**
new ebuild scite-4.1.2.ebuild, old EAPI fixed
Bug: https://bugs.gentoo.org/647586
Closes: https://bugs.gentoo.org/647586
Closes: https://github.com/gentoo/gentoo/pull/10534

Signed-off-by: Ervin Peters <ervin.peters@ervnet.de>
Package-Manager: Portage-2.3.51, Repoman-2.3.11

### CODE REVIEW

File: app-editors/scite/scite-4.1.2.ebuild [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+	#  add the ebuild suffix as shell type for working with ebuilds
+	sed -i "${WORKDIR}/scite/src/perl.properties" \
+		-e "s#\*.sh;\*.bsh;#\*.ebuild;\*.sh;\*.bsh;#" \
+		|| die "error patching /scite/src/perl.prperties"
+
+	# it seems that pwd here is ${S}, but user patches are relative to ${workdir}
+	# Bug #576162
+
+	cd "${WORKDIR}"
```
Comment (monsieurp) [inline app-editors/scite/scite-4.1.2.ebuild]: missing `|| die`

File: app-editors/scite/scite-3.5.5.ebuild [hunk 2]
```diff
@@ -1,7 +1,7 @@
-# Copyright 1999-2018 Gentoo Foundation
+# Copyright 1999-2018 Gentoo Authors
 # Distributed under the terms of the GNU General Public License v2
 
-EAPI="5"
+EAPI="6"
```
Comment (monsieurp) [inline app-editors/scite/scite-3.5.5.ebuild]: Please avoid doing this but instead perform a revision bump.

Comment (juippis) [reply app-editors/scite/scite-3.5.5.ebuild]: (Why are you touching 3.5.5 in the first place? This change seems rather useless)

File: app-editors/scite/scite-4.1.2.ebuild [hunk 3]
```diff
@@ -0,0 +1,112 @@
+# Copyright 1999-2018 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=6
+
+inherit toolchain-funcs flag-o-matic gnome2-utils xdg-utils
```
Comment (juippis) [inline app-editors/scite/scite-4.1.2.ebuild]: You'll have to fix this commit, so you can sort this list alphabetically to improve ebuilds readability.

File: app-editors/scite/scite-4.1.2.ebuild [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+RDEPEND="dev-libs/glib:=
+	x11-libs/cairo
+	x11-libs/gtk+:3=
+	x11-libs/gdk-pixbuf
+	x11-libs/pango
+	lua? ( >=dev-lang/lua-5:= )"
+
+DEPEND="${RDEPEND}
+	virtual/pkgconfig
+	>=sys-devel/gcc-7"
```
Comment (juippis) [inline app-editors/scite/scite-4.1.2.ebuild]: Is **gcc** required, or is 7 required only when gcc is being **used**?

File: app-editors/scite/scite-4.1.2.ebuild [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+
+LICENSE="HPND lua? ( MIT )"
+SLOT="0"
+KEYWORDS="~amd64 ~ppc ~x86 ~x86-fbsd ~amd64-linux ~arm-linux ~x86-linux"
+IUSE="lua"
+
+RDEPEND="dev-libs/glib:=
+	x11-libs/cairo
+	x11-libs/gtk+:3=
+	x11-libs/gdk-pixbuf
```
Comment (juippis) [inline app-editors/scite/scite-4.1.2.ebuild]: Please, sort this list too.

File: app-editors/scite/scite-4.1.2.ebuild [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=6
+
+inherit toolchain-funcs flag-o-matic gnome2-utils xdg-utils
+
+MY_PV=${PV//./}
+DESCRIPTION="A very powerful, highly configurable, small editor with syntax coloring."
+HOMEPAGE="http://www.scintilla.org/SciTE.html"
+SRC_URI="http://www.scintilla.org/${PN}${MY_PV}.tgz"
```
Comment (juippis) [inline app-editors/scite/scite-4.1.2.ebuild]: `-> ${P}.tgz`

File: app-editors/scite/scite-4.1.2.ebuild [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+src_compile() {
+	# prepare make options
+	local emake_pars="GTK3=1"
+	if ! use lua; then
+		emake_pars+=" NO_LUA=1"
+	fi
+
+	emake CC="$(tc-getCC)" LD="$(tc-getLD)" \
+			LDFLAGS="$(raw-ldflags)" AR="$(tc-getAR)" \
+			-C "${WORKDIR}/scintilla/gtk" $emake_pars
```
Comment (juippis) [inline app-editors/scite/scite-4.1.2.ebuild]: Please use `"${...}"` with your variable in here.

File: app-editors/scite/scite-4.1.2.ebuild [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+	emake $emake_pars
```
Comment (juippis) [inline app-editors/scite/scite-4.1.2.ebuild]: Please use `"${...}"` with your variable in here.

File: app-editors/scite/scite-4.1.2.ebuild [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+	emake DESTDIR="${ED}" install
+
+	# we have to keep this because otherwise it'll break upgrading
+	mv "${ED}/usr/bin/SciTE" "${ED}/usr/bin/scite" || die
+	dosym scite /usr/bin/SciTE
+	doman ../doc/scite.1
+	dodoc ../README
+}
+
+pkg_pretend() {
```
Comment (juippis) [inline app-editors/scite/scite-4.1.2.ebuild]: You could sort these phases to match the order they get executed in, 
https://devmanual.gentoo.org/ebuild-writing/functions/index.html

File: app-editors/scite/scite-4.1.2.ebuild [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+	# older gcc is not supported
+	[[ "${MERGE_TYPE}" != "binary" && $(gcc-major-version) -lt 7 ]] && \
+		die "Sorry, but gcc earlier than 7.0 will not work for SCiTE."
+}
+
+pkg_preinst() {
+	gnome2_icon_savelist
```
Comment (juippis) [inline app-editors/scite/scite-4.1.2.ebuild]: (I think this doesn't need to be called anymore, but am not 100 % sure)

File: app-editors/scite/scite-4.1.2.ebuild [hunk 11]
```diff
@@ -0,0 +1,112 @@
+# Copyright 1999-2018 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=6
+
+inherit toolchain-funcs flag-o-matic gnome2-utils xdg-utils
+
+MY_PV=${PV//./}
+DESCRIPTION="A very powerful, highly configurable, small editor with syntax coloring."
+HOMEPAGE="http://www.scintilla.org/SciTE.html"
```
Comment (juippis) [inline app-editors/scite/scite-4.1.2.ebuild]: `https://`

File: app-editors/scite/scite-4.1.2.ebuild (hunk 6 above)
Comment (juippis) [inline app-editors/scite/scite-4.1.2.ebuild]: `https://`

**Comment by juippis** [comment]
Also this is your 3rd PR for this same bump. Whatever it is you're doing wrong, it can **most likely** be fixed with git. Chill a bit and try to learn the tools you're working with, we can try to help too :)

**Comment by dermoench42** [comment]
You're funny guys. find errors and enhancements and did't apply it.
Asking 'Do you wish to keep on as proxied maintainer, since we didn't see any action in 1 month. And suggest to chill out ;) Ok, I shouldn't expect consistency that much, especially in social communication parts ;)

 I would have no problems if you fix that on the fly, because it is just a tool and the changes are cosmetic.

OK I learn.

I learned much the last days, yes, and like to learn much the next time, and I would like to read all this in some resources in developer docs. Since bug reports are rather lazy recognized by developers nowadays, the github PRs seem to be the main option to contribute. Imo it might be a good Idea to show somewhere the livecycle of a PR with some hints to ebuild information. (I love explaining videos and additinal textdetails)

Repoman complained about the EAPI->change to 6->repoman didn't complain anymore ;)

Although the old version before 4 IMHO can be removed, I didn't do that because is is another logical task and it was told me that peoply love small tasks responsible for one topic (KISS).

The part gcc Version Checking is bit more complicated, since there are other compilers around. But how to manage these constraints? It is a logical language feature constraint. How to code them without build test on an arrangement of machines/Compilers?
The author, from the README:
"You must first have GTK+ 2.24 or later and GCC (7.1 or better) installed.
Clang may be used by adding CLANG=1 to the make command line.
Other C++ compilers may work but may require tweaking the make file."

How is it done best?

more questions:

Is there a way to repair this PR or should I go to the next cycle?

Could I remove the 3.x ebuild and what checks should I do before?

Could I put this together in one PR?

Ervin

**Comment by juippis** [comment]
>Asking 'Do you wish to keep on as proxied maintainer, since we didn't see any action in 1 month. And suggest to chill out ;) Ok, I shouldn't expect consistency that much, especially in social communication parts ;)

First of all, sorry, I wasn't aware of your 'history'. I didn't mean 'chill' as an insult, I meant you already received feedback on your first PR, and closing old and opening new PRs loses this feedback and makes tracking progress harder. Also, your PR gets thrown back of the review queue again, and makes it longer and longer to get it merged. So let me answer this first: 

>Is there a way to repair this PR or should I go to the next cycle?

Yes. You can most likely fix whatever you've done by using `git rebase`. If that doesn't work, save your work somewhere temporary and do a hard reset. Or just copypaste from this PR before pushing over. Then start over in your local branch, and force-push your work. There are the keywords that help you  do research and practice.

>I would have no problems if you fix that on the fly, because it is just a tool and the changes are cosmetic.

The thing is, if you want to be a proxy-maintainer, you need to have required skills and knowledge to work as one. The proxy-maint team isn't here to rewrite your contributions, but to guide you getting your contribution to match Gentoo standards. Yes, sometimes some small things can be fixed while merging because it's pointless to longen the merge time and stress CI bot over a small typo or something. But **to my eyes**, this has more work and even some build-breaking stuff to fix for. 

>Imo it might be a good Idea to show somewhere the livecycle of a PR with some hints to ebuild information. (I love explaining videos and additinal textdetails)

I'm not sure if I got this right, but devmanual is a nice place to get info about what is expected, altough you don't really need to read the whole thing for one contribution.
https://devmanual.gentoo.org/

>Repoman complained about the EAPI->change to 6->repoman didn't complain anymore ;)

Right, I can see this happening. You should only care about what issues your changes create, not what are present already. Yes, EAPI-5 is deprecated and will be rid of in time. Also, in future, you can't make changes to stable ebuilds AND EAPI-bump itself requires a revbump to ebuilds. 
https://devmanual.gentoo.org/general-concepts/ebuild-revisions/index.html
https://wiki.gentoo.org/wiki/Proxied_Maintainer_FAQ

>How is it done best?

Now I believe there are multiple ways of doing this. What I'd personally do, is add clang USE flag and force clang being used if that USE is enabled. thunderbird, firefox, compiler-rt and compiler-rt-sanitizers have a nice way of accomplishing this, to name a few packages you can look example from. Remember to append CLANG=1 to either emake, or makefile. Also in that regard, I see multiple repetitions on makefile and emake. Is modifying makefile necessary, if emake passes these options?

As to gcc, the DEPEND line is not really dependable. Users can have 7 installed, but choose to use 6 instead. You'll want to use toolchain-funcs.eclass to satisfy that. @hsoft gave you advice in the bug report linked here, how to achieve that (or where to look example). Note: GCC-7 is stabilized on every arch, so everyone **should** have it installed.

>Could I remove the 3.x ebuild and what checks should I do before?

I see 4.0.2 is stabilized on all arches already. Is there a case where any user would have earlier versions installed? If some prefer the 3xx series then leave it, but if you believe everyone wants 4 instead, you can remove the older ones. Make sure no package in the tree depends on that before removing. (Note: leaving old software to tree can introduce them to different kind of build & runtime bugs due to library mismatch etc, do you want to support these old packages as a proxy-maintainer?)

>Could I put this together in one PR?

Yes. First comes the new version bump, then removal of old ebuilds. 2 separate commits in that order.

Lot of text here. I gave you some reading and research to do, but if it feels like nothing is still making sense, don't be afraid to ask. I'll be more specific :)

**Comment by dermoench42** [comment]
Am Mon, 03 Dec 2018 03:00:41 -0800 schrieb juippis
<notifications@github.com>:

> >Is there a way to repair this PR or should I go to the next cycle?
> 
> Yes. You can most likely fix whatever you've done by using `git
> rebase`. If that doesn't work, save your work somewhere temporary and
> do a hard reset. Or just copypaste from this PR before pushing over.
> Then start over in your local branch, and force-push your work. There
> are the keywords that help you  do research and practice.

got the upstream branch as described in
https://wiki.gentoo.org/wiki/GitHub_Pull_Requests .

got the scite branch from which I generated this PR.

now I do:

$ git checkout master
$ git branch -d scite
$ git pull upstream
$ git branch scite
$ git checkout scite
$ cd ~/git/gentoo/app-editors/scite/
$ cp /usr/local/portage/app-editors/scite/scite-4.1.2.ebuild .
$ repoman manifest
$ repoman -dx full
$ git add Manifest
$ git add scite-4.1.2.ebuild

now I do a 
$ repoman ci 

and see what happens.


> I'm not sure if I got this right, but devmanual is a nice place to
> get info about what is expected, altough you don't really need to
> read the whole thing for one contribution.
> https://devmanual.gentoo.org/

love that doc ;)

> https://devmanual.gentoo.org/general-concepts/ebuild-revisions/index.html

OK.

> https://wiki.gentoo.org/wiki/Proxied_Maintainer_FAQ

> As to gcc, the DEPEND line is not really dependable. Users can have 7
> installed, but choose to use 6 instead. You'll want to use
> toolchain-funcs.eclass to satisfy that. @hsoft gave you advice in the
> bug report linked here,

It was already implemented in pkg_pretend() :))

> how to achieve that (or where to look
> example). Note: GCC-7 is stabilized on every arch, so everyone
> **should** have it installed.

That means, there is no need to support clang in ebuild now.

I'll try again, first the new ebuild with sorted variable contents and
the missing die statement to handle vanishing ${workdir}.

Then I'' propose the cleanout.

ervin

-- 
Ervin Peters
Herderplatz 12
D 99423 Weimar

Mobile: +49 172 2043926

http://coder.ervnet.de

**Comment by dermoench42** [comment]
Seems that something went wrong, git requested a pull and rebase. closing and retry from scratch.

**Pull request closed without merging**